src/v/config/constraint_method_generators.cc

```diff
@@ -0,0 +1,561 @@
+          args.name);
+        topic_val = *min_max;
+    }
+}
+
+template<typename T>
+bool match_cluster_property(
+  const T& topic_val,
+  const T& cluster_val,
+  const std::variant<
+    range_values<int64_t>,
+    range_values<uint64_t>,
+    constraint_enabled_t>& flags) {
+    return ss::visit(
+      flags,
+      [](const range_values<int64_t>) { return false; },
+      [](const range_values<uint64_t>) { return false; },
+      [&topic_val, &cluster_val](const constraint_enabled_t enabled) {
+          return enabled == constraint_enabled_t::yes
+                 && topic_val == cluster_val;
+      });
+}
+
+template<typename T>
+bool validate_property(
+  const T& topic_val,
+  const T& cluster_val,
+  const model::topic& topic,
+  const constraint_args& args) {
+    if (!match_cluster_property(topic_val, cluster_val, args.flags)) {
+        logger.error(
+          "Constraints failure[does not match the cluster default {}]: "
+          "topic property {}.{}, value {}",
+          cluster_val,
+          topic(),
+          args.name,
+          topic_val);
+        return false;
+    }
+
+    return true;
+}
+
+template<typename T>
+void maybe_clamp_property(
+  T& topic_val,
+  const T& cluster_val,
+  const model::topic& topic,
+  const constraint_args& args) {
+    if (!match_cluster_property(topic_val, cluster_val, args.flags)) {
+        logger.warn(
+          "Overwriting topic property {}.{} to the cluster default {}",
+          topic(),
+          args.name,
+          cluster_val);
+        topic_val = cluster_val;
+    }
+}
+
+template<typename TopicFunc, typename ClusterFunc>
+constraint_validator_t
+make_validator(TopicFunc&& get_topic_val, ClusterFunc&& get_cluster_val) {
+    return [get_topic_val{std::forward<TopicFunc>(get_topic_val)},
+            get_cluster_val{std::forward<ClusterFunc>(get_cluster_val)}](
+             const cluster::topic_configuration& topic_cfg,
+             const constraint_args& args) {
+        auto topic_val = get_topic_val(topic_cfg);
+        return ternary_property_op(
+          config::shard_local_cfg().get(args.name),
+          [&topic_val, &topic_cfg, &args] {
+              return validate_range<int64_t>(
+                topic_val, topic_cfg.tp_ns.tp, args);
+          },
+          [&topic_val, &topic_cfg, &args] {
+              return validate_range<uint64_t>(
+                topic_val, topic_cfg.tp_ns.tp, args);
+          },
+          [&topic_val, &topic_cfg, &args, &get_cluster_val] {
+              // For property with enabled constraint
+              return validate_property(
+                topic_val, get_cluster_val(), topic_cfg.tp_ns.tp, args);
+          });
+    };
+}
+
+template<typename TopicFunc>
+constraint_validator_t make_validator(TopicFunc&& get_topic_val) {
+    return make_validator(
+      get_topic_val, [get_cluster_val{std::forward<TopicFunc>(get_topic_val)}] {
+          return get_cluster_val(cluster::topic_configuration{});
+      });
+}
```

Comment:
I'm a little lost trying to parse this out.  In which situations would you use the `make_validator`? This is to lock a topic config to a cluster config, right?

Reply:
I think it's the parameter naming that's throwing me here

Reply:
Removed.